### PR TITLE
Adding proper OpenTelemetry integration via. registration helpers and better context propagation

### PR DESCRIPTION
**Pull request by stebet**
## Proposed Changes

* Making it possible to override the default context propagation in RabbitMQActivitySource to make it possible to have proper OpenTelemetry Baggage propagation
* Adding a separate OpenTelemetry integration package to make it easier to integrate with OTel and configure the RabbitMQActivitySource
* Adding new tests for OTel integration package and context propagation

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [x] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [X] I have read the `CONTRIBUTING.md` document
- [X] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [X] All tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

Would love to get some input from @lmolkova and @martinjt as well :)

### CODE REVIEW

**Comment by lukebakken** [comment]
@stebet we can leave this PR as a Draft until it's ready.

**Comment by stebet** [comment]
@lukebakken I'm thinking I might just move the OTel integration tests to the SequentialIntegration test project. They are pretty much doing a very similar thing to the ActivitySource integration tests and just referencing an extra project (not drastically different like the OAuth2 tests). Does that make sense to not convolute the test projects and pipelines?

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+namespace OpenTelemetry.Trace
+{
+    public static class OpenTelemetryExtensions
+    {
+        internal static TextMapPropagator s_propagator = Propagators.DefaultTextMapPropagator;
+
+        public static TracerProviderBuilder AddRabbitMQ(this TracerProviderBuilder builder,
+            RabbitMQOpenTelemetryConfiguration configuration)
+        {
+            if (configuration.PropagateBaggage)
```
Comment (lmolkova) [inline projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:19]: any reason not to allow user to provide a preconfigured propagator instance?
E.g. users could want to use [B3 propagation](https://github.com/openzipkin/b3-propagation) or [XRay AWS propagator](https://github.com/aws-observability/aws-otel-dotnet)

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                {
+                    new TraceContextPropagator(), new BaggagePropagator()
+                });
+            }
+            else
+            {
+                s_propagator = new TraceContextPropagator();
+            }
+
+            RabbitMQActivitySource.UseRoutingKeyAsOperationName = configuration.UseRoutingKeyAsOperationName;
```
Comment (lmolkova) [inline projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:31]: do you feel it's something we should document better /provide more clear guidance on in the messaging semconv?

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 1 above)
Comment (stebet) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:19]: No reason bu my lack of knowledge :). Are you then referring to these as alternatives to the tracestate propagator? Are there examples for me to look at?

Comment (lmolkova) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:19]: I mean you're forcing a specific propagator - a `CompositeTextMapPropagator` or `TraceContextPropagator` which is not necessary.

I believe you should just use the `DefaultTextMapPropagator`. 
Users can configure it however they want to, check out  https://github.com/open-telemetry/opentelemetry-dotnet/blob/9b246750ab76e94986fc433d5c7af046a137e477/src/OpenTelemetry.Extensions.Propagators/README.md?plain=1#L29

If I understand the intent correctly you want to users to be able to disable baggage propagation - any specific reason?

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            RabbitMQActivitySource.ContextExtractor = OpenTelemetryContextExtractor;
+            RabbitMQActivitySource.ContextInjector = OpenTelemetryContextInjector;
+
+            if (configuration.IncludeSubscribers)
```
Comment (lmolkova) [inline projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:35]: I don't have any strong opinion here, but want to challenge the need for this flag (and `IncludePublishers`)
if users want more control, they can always call `AddSource` themselves.

I guess my broader question is when would users want to enable rabbitmq via an instrumentation library if all they really need to do is call
`builder.AddSource("RabbitMQ.Client.*")`

**Comment by danielmarbach** [comment]
@stebet thanks for the ping. Currently on vacation and after that pretty occupied with some other things. I'm pinging @lailabougria for a review after her vacation. She has lots of knowledge in this area and it quite likely the better reviewer for this type of work anyway

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 1 above)
Comment (stebet) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:19]: > I mean you're forcing a specific propagator - a `CompositeTextMapPropagator` or `TraceContextPropagator` which is not necessary.
> 
> 
> 
> I believe you should just use the `DefaultTextMapPropagator`. 
> 
> Users can configure it however they want to, check out  https://github.com/open-telemetry/opentelemetry-dotnet/blob/9b246750ab76e94986fc433d5c7af046a137e477/src/OpenTelemetry.Extensions.Propagators/README.md?plain=1#L29

Ahh gotcha! Saw that the DefaultTextMapPropagator was a Noop so I thought this was the way. I'll change it :)
> 
> 
> If I understand the intent correctly you want to users to be able to disable baggage propagation - any specific reason?

Nope. I'll remove it. Does it make sense to then have a composite propagator with the default textmap propagator and the baggage propagator or do users normally configure the baggage propagator themselves?

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 2 above)
Comment (stebet) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:31]: Might be a good idea if people are running into high-cardinality issues with their messaging span names.

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 3 above)
Comment (stebet) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:35]: > I don't have any strong opinion here, but want to challenge the need for this flag (and `IncludePublishers`) if users want more control, they can always call `AddSource` themselves.
> 
> I guess my broader question is when would users want to enable rabbitmq via an instrumentation library if all they really need to do is call `builder.AddSource("RabbitMQ.Client.*")`

The main reason for the library was to simplify the OTel configuration and setup for users (with the extension method) and to work around the current issue that since OTel and Activity Baggage propagation is different we do not have to make the `OpenTelemetry.Api` package a dependency on the RabbitMQ library itself.

**Comment by stebet** [comment]
@lukebakken I think this PR is ready for a proper review now.

Thanks for the feedback @lmolkova and @martinjt, feel free to provide additional comments and feedback if you have time to spare :)

CCing @davidfowl in case he wants to take a look as well from an Aspire perspective.

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 2 above)
Comment (martinjt) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:22]: Is Operation the right name here? Since this is focused on OpenTelemetry, I'm wondering if just "SpanName" is a better wording.

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 3 above)
Comment (martinjt) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:26]: I'm honestly not sure whether the separation of allowing people to turn on/off subscriber or publisher. Is that actually a usecase?

I can understand adding it here if that's a usecase seeing as this is for simplification. My question is, should you allow people to just do one or the other?

Comment (stebet) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:26]: > I'm honestly not sure whether the separation of allowing people to turn on/off subscriber or publisher. Is that actually a usecase?
> 
> I can understand adding it here if that's a usecase seeing as this is for simplification. My question is, should you allow people to just do one or the other?

No idea, but I think it was you and @lmolkova that suggested I separated the ActivitySources in the earlier OTel pr: https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1261/files#r1392469382

I have no preference either way really

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 2 above)
Comment (stebet) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:22]: OTel config: `UseRoutingKeyInSpanName`
RabbitMQActivitySource property: `UseRoutingKeyInActivityName`
Make more sense?

**Comment by stebet** [comment]
This PR should also help with #1478

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 1 above)
Comment (lmolkova) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs]: The otel sdk should set the default one to the [composite](https://github.com/open-telemetry/opentelemetry-specification/blob/a03382ada8afa9415266a84dafac0510ec8c160f/specification/context/api-propagators.md?plain=1#L310).

Not completely sure what sets `OpenTelemetry.Context.Propagation.Propagators.DefaultTextMapPropagator`, but I see it's set after OTel SDK is configured:)

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 2 above)
Comment (lmolkova) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:22]: > Might be a good idea if people are running into high-cardinality issues with their messaging span names.

the span name cardinality requirements are soft (~undefined~) - https://github.com/open-telemetry/opentelemetry-specification/issues/3534. Things like HTTP route we put there are like middlish-cardinality and I think the same is true about queue names/routing key.

so if cardinality is the only problem, I'd not make span name configurable (worst case users can rename spans with a processor). So again, no strong opinions

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 3 above)
Comment (lmolkova) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:26]: Having different names is more future-proof, but also it costs nothing since you can enable sources with wildcard.

I didn't try to push you to remove the instrumentation library, was just trying to understand what'd be the better case for it.

The `Activity.Baggage` problem is an  interesting one. @martinjt do you know if there are any plans to make baggages work better together? I can create issues/ping people/etc - it'd be great if we could solve it for everyone without instrumentations trying to patch it in each case.

Comment (stebet) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:26]: > The `Activity.Baggage` problem is an interesting one. @martinjt do you know if there are any plans to make baggages work better together? I can create issues/ping people/etc - it'd be great if we could solve it for everyone without instrumentations trying to patch it in each case.

Here's an old issue that relates to it (CC @cijothomas): https://github.com/open-telemetry/opentelemetry-dotnet/issues/1842

**Comment by stebet** [comment]
Think I've resolved most of the comments @lukebakken :)

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs (hunk 3 above)
Comment (cijothomas) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs]: The Activity baggage vs OTel Baggage is still an unresolved problem! It'll eventually be fixed, though we don't have specifics or timeline. (its not likely coming in this year, there is nothing tracking it in .NET 9 release timeframe.)

File: projects/RabbitMQ.Client.OpenTelemetry/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+// Configure the OpenTelemetry SDK to trace ASP.NET Core, the RabbitMQ .NET Client and export the traces to the console.
+// Also configures context propagation to propagate the TraceContext and Baggage using the W3C specification.
+
+var compositeTextMapPropagator = new CompositeTextMapPropagator(new TextMapPropagator[]
+{
+    new TraceContextPropagator(),
+    new BaggagePropagator()
+});
+
+Sdk.SetDefaultTextMapPropagator(compositeTextMapPropagator);
```
Comment (cijothomas) [inline projects/RabbitMQ.Client.OpenTelemetry/README.md:48]: this is already the default.... so why is this needed again?

Comment (stebet) [reply projects/RabbitMQ.Client.OpenTelemetry/README.md:48]: When I was running the tests it seemed like the NoopPropagator was the default. I'll take another look.

**Comment by lailabougria** [comment]
Thanks, @danielmarbach, for the ping!

Regarding @martinjt's [comment](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1528#discussion_r1549378766):
> I'm honestly not sure whether the separation of allowing people to turn on/off subscriber or publisher. Is that actually a usecase?

I believe splitting the ActivitySources is ok even though it may not really be necessary eventually; having them split does provide more flexibility and possibly avoids breaking changes down the road.

However, what stood out to me is that the user can't easily benefit from those separate ActivitySources with the proposed instrumentation API, as there's no way to configure them separately and [the API adds both sources by default](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1528/files#diff-76c665a7f89ccce6852ccadf26a003224f84ebcf00c57e461af95b301cac639dR18).  Assuming users will use the instrumentation API `AddRabbitMQInstrumentation`, instead of `AddSource`, they can't opt-in to individual activity sources that way, so they'd have to purposefully use `AddSource` instead of the instrumentation API, which also means the other code in that method won't be invoked (meaning the baggage propagation issue is still at play, and the propagators aren't set automatically, as @stebet explained in this [comment](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1528#discussion_r1549256549)). 

That may be something to consider in the design of the instrumentation APIs if the decision is to keep the ActivitySources split.

File: projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace OpenTelemetry.Trace
+{
+    public static class OpenTelemetryExtensions
+    {
+        public static TracerProviderBuilder AddRabbitMQInstrumentation(this TracerProviderBuilder builder)
+        {
+            RabbitMQActivitySource.UseRoutingKeyAsOperationName = true;
+            RabbitMQActivitySource.ContextExtractor = OpenTelemetryContextExtractor;
+            RabbitMQActivitySource.ContextInjector = OpenTelemetryContextInjector;
```
Comment (Kielek) [inline projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:17]: Why we need this part? Can it be set by default in main package?

I would expect that this method will only call `builder.AddSource("RabbitMQ.Client.*")` as a convenient way to register it in TracerProviderBuilder. Other parts should be done in the instrumentation library.

Alternative options is to have here the second method:

```csharp
public static TracerProviderBuilder AddRabbitMQInstrumentation(this TracerProviderBuilder builder, Action<RabbitMQSpecificOptions>? configure);
```

where you can configure other methods externally.

Comment (stebet) [reply projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs:17]: OpenTelemetryContextInjector is OTel specific for Baggage handling, that's pretty much the only reason for the package, because we don't want to pull OpenTelemetry.API dependencies into the main library.

**Comment by stebet** [comment]
> Thanks, @danielmarbach, for the ping!
> 
> Regarding @martinjt's [comment](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1528#discussion_r1549378766):
> 
> > I'm honestly not sure whether the separation of allowing people to turn on/off subscriber or publisher. Is that actually a usecase?
> 
> I believe splitting the ActivitySources is ok even though it may not really be necessary eventually; having them split does provide more flexibility and possibly avoids breaking changes down the road.
> 
> However, what stood out to me is that the user can't easily benefit from those separate ActivitySources with the proposed instrumentation API, as there's no way to configure them separately and [the API adds both sources by default](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1528/files#diff-76c665a7f89ccce6852ccadf26a003224f84ebcf00c57e461af95b301cac639dR18). Assuming users will use the instrumentation API `AddRabbitMQInstrumentation`, instead of `AddSource`, they can't opt-in to individual activity sources that way, so they'd have to purposefully use `AddSource` instead of the instrumentation API, which also means the other code in that method won't be invoked (meaning the baggage propagation issue is still at play, and the propagators aren't set automatically, as @stebet explained in this [comment](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1528#discussion_r1549256549)).
> 
> That may be something to consider in the design of the instrumentation APIs if the decision is to keep the ActivitySources split.

We discussed this exact thing here: https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1528#discussion_r1548793930

The consensus seemed to be that it didn't really make sense from the end-user perspective.

**Comment by lukebakken** [comment]
@stebet I merged `main` into your branch to resolve conflicts, but I had to make this change, which I'm guessing may not be correct:

https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1528/files#diff-f6d8753c982a8409794d13d5a9263aea70541146d5f4255c86e4e5150f6b9a1dR1072-R1075

Merging #1233 will probably address the above, however.

File: projects/RabbitMQ.Client.OpenTelemetry/RabbitMQ.Client.OpenTelemetry.csproj [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+  <PropertyGroup>
+    <TargetFrameworks>net6.0;netstandard2.0</TargetFrameworks>
+    <NoWarn>$(NoWarn);CS1591</NoWarn>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <AssemblyTitle>RabbitMQ OpenTelemetry Integration Package for .NET</AssemblyTitle>
+    <Authors>VMware</Authors>
+    <Company>VMware, Inc. or its affiliates.</Company>
+    <Copyright>Copyright © 2007-2023 VMware, Inc. or its affiliates.</Copyright>
+    <Description>The RabbitMQ OAuth2 Client Library for .NET enables OAuth2 token refresh for RabbitMQ.Client</Description>
```
Comment (WeihanLi) [inline projects/RabbitMQ.Client.OpenTelemetry/RabbitMQ.Client.OpenTelemetry.csproj:11]: Seemed this description need to be updated

File: projects/RabbitMQ.Client.OpenTelemetry/RabbitMQ.Client.OpenTelemetry.csproj [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    <IncludeSymbols>true</IncludeSymbols>
+    <SymbolPackageFormat>snupkg</SymbolPackageFormat>
+    <AssemblyOriginatorKeyFile>../rabbit.snk</AssemblyOriginatorKeyFile>
+    <SignAssembly>true</SignAssembly>
+    <MinVerTagPrefix>otel-</MinVerTagPrefix>
+    <MinVerVerbosity>minimal</MinVerVerbosity>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
+    <PackageOutputPath>../../packages</PackageOutputPath>
+    <PackageReadmeFile>README.md</PackageReadmeFile>
+    <LangVersion>7.3</LangVersion>
```
Comment (WeihanLi) [inline projects/RabbitMQ.Client.OpenTelemetry/RabbitMQ.Client.OpenTelemetry.csproj:29]: Any reason stick to `7.3`?

Comment (danielmarbach) [reply projects/RabbitMQ.Client.OpenTelemetry/RabbitMQ.Client.OpenTelemetry.csproj:29]: I would suggest at least 9 to have static lambda support

Comment (lukebakken) [reply projects/RabbitMQ.Client.OpenTelemetry/RabbitMQ.Client.OpenTelemetry.csproj:29]: https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1424

File: projects/RabbitMQ.Client.OpenTelemetry/RabbitMQ.Client.OpenTelemetry.csproj (hunk 6 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client.OpenTelemetry/RabbitMQ.Client.OpenTelemetry.csproj:11]: Thank you for noticing!

**Comment by lukebakken** [comment]
Just FYI for everyone who has commented and contributed here, I would like to release version 7 this month (May 2024).

**Comment by lukebakken** [comment]
Hi everyone -

In the interest of shipping v7 of this library this month, I am going to merge this PR as-is today or tomorrow. I am going to release RC1 of version 7 by the end-of-day tomorrow, June 4th, 2024.

I am assuming the work here is complete. As I've said in other OpenTelemetry-related PRs, I'm not very familiar with OTel and rely on others' expertise.

**Review by lukebakken — APPROVED** [review]
Thank you @stebet
